restler/utils/logging/trace_db.py

```diff
@@ -0,0 +1,225 @@
+        only one logger at a time is supported).
+        The default format is newline-delimited json.
+    """
+    def __init__(self, storage_writer):
+        self.storage_writer = storage_writer
+        self._log_queue = queue.SimpleQueue()
+        self._finished = False
+
+    def log(self, message):
+        if self._finished:
+            raise Exception("ERROR: log() should not be called after finish().")
+        self._log_queue.put(message)
+
+    def log_queue_empty(self):
+        return self._log_queue.empty()
+
+    def finish(self):
+        if self._finished:
+            raise Exception("ERROR: finish() should only be called once.")
+        self._finished = True
+        self._log_queue.put(None)
+
+    def save_log_message(self):
+        """Takes a message off the queue and saves it using the specified serializer.
+            If fewer messages are available, it will save all available messages.
+        """
+        message = self._log_queue.get()
+        if message is None:
+            # End of tracing signal
+            return
+        self.storage_writer.save(message)
+
+    def get_sequence_id(self):
+        tls_trace_log = get_trace_log()
+        if tls_trace_log is None:
+            return None
+        return tls_trace_log.sequence_id
+
+    def log_request_response(self, request=None, response=None,  tags={}, timestamp=None):
+        if request is None and response is None:
+            raise Exception("ERROR: Request or response must be specified.")
+        try:
+            tls_trace_log = get_trace_log()
+            if tls_trace_log is None:
+                # Logging requests outside of sequence context (e.g. GC or checker requests
+                # that are not part of a sequence)
+                trace_log = RequestTraceLog()
+            else:
+                trace_log = RequestTraceLog(request_id=tls_trace_log.request_id,
+                                              sequence_id=tls_trace_log.sequence_id,
+                                              combination_id=tls_trace_log.combination_id,
+                                              tags=tls_trace_log.tags)
+            if request:
+                trace_log.request = request
+                if timestamp is not None:
+                    trace_log.sent_timestamp = timestamp
+            if response:
+                trace_log.response = response
+                if timestamp is not None:
+                    trace_log.received_timestamp = timestamp
+            if tags:
+                trace_log.tags.extend(tags)
+            self.log(trace_log.to_dict())
+        except Exception as error:
+            # print the callstack
+            import traceback
+            traceback.print_exc()
+            print(f"Warning: Exception logging request/response: {error}")
+            pass
+
+    def load_trace_data(self):
+        self.storage_writer.load()
+
+    def normalize_trace_data(self):
```

Comment:
Is the `normalize_trace_data` method used anywhere today, or is this intended to be called once the TODO in ndjson serializer's `normalize()` implementation is complete?

(Asking because I couldn't find where this method is being used, and I wasn't sure if I missed it)

Reply:
Correct, it is not used yet.  Will add a TODO.